pkgs/stdenv/generic/make-derivation.nix

```diff
@@ -91,6 +91,7 @@ in rec {
 
     , ... } @ attrs:
 
+      assert !lib.hasAttr "name" attrs -> lib.assertMsg (lib.hasAttrs [ "pname" "version" ] attrs ) "Derivation needs a pname and a version";
```

Comment:
assertions tend to break overriding. Did you test whether this functions?

Reply:
I've made quick test with this 
```nix
{ pkgs, ... }:
{
  doas-no-pam = pkgs.doas.override { withPAM = false; };
  pgpdump-no-compressed-packets = pkgs.pgpdump.override { supportCompressedPackets = false; };
  stoken-no-gtk3 = pkgs.stoken.override { withGTK3 = false; };
  tpm2-tools-no-abrmd = pkgs.tpm2-tools.override { abrmdSupport = false; };
}
```

And it seems to work fine.
(all these derivations use stdenv.mkDerivation)